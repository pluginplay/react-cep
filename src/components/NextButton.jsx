import React from 'react'
import Button from './Button'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import Flex from 'styled-flex-component'
import arrowRight from '../assets/icons/arrow-right.svg'

class NextButton extends React.Component {
  render () {
    return (
      <Button onClick={this.props.onClick} className={this.props.className}>
        <Flex full alignCenter>
          <span>{this.props.children}</span>
          <img src={arrowRight} alt={'Next'} />
        </Flex>
      </Button>
    )
  }
}

NextButton.propTypes = {
  onClick: PropTypes.func.isRequired
}

export default styled(NextButton)`
  background: ${props => props.theme.ae.systemHighlightColor};
  border: none;
  span {
    margin-right: 5px;
  }
  img {
    height: 12px;
    margin-left: auto;
  }
  &:hover, &:active, &:focus {
    background: ${props => props.theme.ae.systemHighlightColor};
  }
  &:hover {
    background: ${props => props.theme.background.darker};
  }
`
