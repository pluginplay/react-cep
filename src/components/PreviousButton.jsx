import React from 'react'
import Button from './Button'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import Flex from 'styled-flex-component'
import arrowLeft from '../assets/icons/arrow-left.svg'

class PreviousButton extends React.Component {
  render () {
    return (
      <Button onClick={this.props.onClick} className={this.props.className}>
        <Flex full alignCenter>
          <img src={arrowLeft} alt={'Back'} />
          <span>Back</span>
        </Flex>
      </Button>
    )
  }
}

PreviousButton.propTypes = {
  onClick: PropTypes.func.isRequired
}

export default styled(PreviousButton)`
  background: none;
  border: none;
  min-width: 0;
  img {
    height: 12px;
    margin-right: 5px;
  }
  &:active, &:focus {
    background: none;    
  }
`
