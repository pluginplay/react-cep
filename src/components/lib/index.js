import SystemState from '../../models/SystemState'
import PersistedState from '../../models/PersistedState'

export { SystemState, PersistedState }
