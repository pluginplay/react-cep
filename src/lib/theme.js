import * as csInterface from 'cep-lib/csinterface'
import { lighten, darken } from 'polished'

export const getDefaultTheme = () => {
  const ae = csInterface.getThemeInformation()
  ae.systemHighlightColor = '#0078d7' // We have to hard-code this because Windows gets the wrong color.
  return {
    ae,
    background: {
      lighter: lighten(0.2, ae.panelBackgroundColor),
      light: lighten(0.1, ae.panelBackgroundColor),
      dark: darken(0.05, ae.panelBackgroundColor),
      darker: darken(0.1, ae.panelBackgroundColor)
    },
    colors: {
      green: '#23EB87',
      red: '#eb093f',
      blue: ae.systemHighlightColor
    }
  }
}
