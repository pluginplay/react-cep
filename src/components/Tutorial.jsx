import React from 'react'
import styled from 'styled-components'
import Flex from 'styled-flex-component'
import NextButton from './NextButton'
import PreviousButton from './PreviousButton'
import PropTypes from 'prop-types'

class Tutorial extends React.Component {
  constructor (props) {
    super(props)
    this.state = { currentStep: 0 }
    this.onNextStepClickedBound = this.onNextStepClicked.bind(this)
    this.onPreviousStepClickedBound = this.onPreviousStepClicked.bind(this)
    this.onFinalButtonClickedBound = this.onFinalButtonClicked.bind(this)
  }

  get _currentStep () { return this.props.steps[this.state.currentStep] }
  get _shouldShowNext () { return this.state.currentStep < this.props.steps.length - 1 }
  get _shouldShowFinal () { return this.state.currentStep === this.props.steps.length - 1 }
  get _shouldShowPrevious () { return this.state.currentStep > 0 }

  onFinalButtonClicked () {
    if (this.props.onFinalButtonClicked) {
      this.props.onFinalButtonClicked()
    }
  }

  onNextStepClicked () {
    this.setState({ currentStep: this.state.currentStep + 1 })
  }

  onPreviousStepClicked () {
    this.setState({ currentStep: this.state.currentStep - 1 })
  }

  renderFinalButton () {
    if (this._shouldShowFinal) {
      return (
        <NextButton className={'next-button final-button'} onClick={this.onFinalButtonClickedBound}>
          {this.props.finalButtonText}
        </NextButton>
      )
    } else return null
  }

  renderNextButton (text) {
    if (this._shouldShowNext) {
      return (
        <NextButton className={'next-button'} onClick={this.onNextStepClickedBound}>{text}</NextButton>
      )
    } else return null
  }

  renderPreviousButton () {
    if (this._shouldShowPrevious) {
      return (
        <PreviousButton onClick={this.onPreviousStepClickedBound} className={'previous-button'} />
      )
    } else return null
  }

  render() {
    return (
      <Flex full justifyCenter className={this.props.className}>
        <div className={'side image-side'}>
          <Flex full center className={'image-container'}>
            <img src={this._currentStep.image} />
          </Flex>
        </div>
        <div className={'side copy-side'}>
          <Flex full column justifyCenter>
            <h2>{this._currentStep.title}</h2>
            {this._currentStep.description()}
            <Flex alignCenter>
              {this.renderPreviousButton()}
              {this.renderNextButton(this._currentStep.nextButton)}
              {this.renderFinalButton()}
            </Flex>
          </Flex>
        </div>
      </Flex>
    )
  }
}

Tutorial.propTypes = {
  finalButtonText: PropTypes.string.isRequired,
  onFinalButtonClicked: PropTypes.func,
  steps: PropTypes.arrayOf(PropTypes.object).isRequired
}

export default styled(Tutorial)`
  flex-grow: 1;
  position: absolute;
  top: 0;
  bottom: 0;
  .side {
    width: 50%;
    height: 100%;
    box-sizing: border-box;
    &.image-side {
      background: ${props => props.theme.background.dark};
      .image-container {
        padding: 15px;
        box-sizing: border-box;
      }
      img {
        max-width: 100%;
        max-height: 100%;
      }
    }
    &.copy-side {
      padding: 30px;
    }
  }
  .previous-button {
    margin-right: auto;
    margin-left: -8px;
  }
  .next-button {
    margin-right: 0;
  }
  .final-button {
    font-weight: bold;
    font-size: 14px;
    span {
      margin-right: 10px;
    }
    img {
      height: 14px;
    }
  }
  h2 { 
    margin: 0;
  }
  a {
    color: white;
  }
`
