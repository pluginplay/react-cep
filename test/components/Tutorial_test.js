import React from 'react'
import renderer from 'react-test-renderer'
import { expect, withTheme } from '../setup'
import { getDefaultTheme } from '../../src/lib/theme'
import { shallow } from 'enzyme'
import Tutorial from '../../src/components/Tutorial'

import PreviousButton from '../../src/components/PreviousButton'
import NextButton from '../../src/components/NextButton'

let tutorial = null
beforeEach(() => {
  tutorial = [
    { title: 'Step 1', image: 'foo', nextButton: 'Step 1 Next', description: () => <p>Step 1</p> },
    { title: 'Step 2', image: 'bar', nextButton: 'Step 2 Next', description: () => <p>Step 2</p> },
    { title: 'Step 3', image: 'par', nextButton: 'Step 3 Next', description: () => <p>Step 3</p> }
  ]
})

const getWrapper = (props = {}) => {
  return shallow(<Tutorial theme={getDefaultTheme()} steps={tutorial} finalButtonText={'Final'} {...props} />).dive()
}

it('renders correctly', () => {
  const tree = renderer
    .create(withTheme(<Tutorial steps={tutorial} finalButtonText={'Final'} />))
    .toJSON()
  jestExpect(tree).toMatchSnapshot()
})

it('shows the previous button only when there is a previous step', () => {
  const wrapper = getWrapper()
  expect(wrapper.find(PreviousButton)).to.have.length(0)
  wrapper.setState({ currentStep: 1 })
  expect(wrapper.find(PreviousButton)).to.have.length(1)
})

it('shows the final button when we are on the final step', () => {
  const wrapper = getWrapper()
  expect(wrapper.find(NextButton).filter('.next-button').not('.final-button')).to.have.length(1)
  expect(wrapper.find(NextButton).filter('.final-button')).to.have.length(0)
  wrapper.setState({ currentStep: 2 })
  expect(wrapper.find(NextButton).filter('.next-button').not('.final-button')).to.have.length(0)
  expect(wrapper.find(NextButton).filter('.final-button')).to.have.length(1)
})

it('shows the final button with the proper text configured', () => {
  const wrapper = getWrapper({ finalButtonText: 'final button text' }).setState({ currentStep: 2 })
  expect(wrapper.find(NextButton).filter('.final-button').first().childAt(0)).to.have.text('final button text')
})

it('shows the next button with the proper text configured', () => {
  const wrapper = getWrapper()
  expect(wrapper.find(NextButton).not('.final-button').first().childAt(0)).to.have.text('Step 1 Next')
})

it('transitions to the next step when next is clicked', () => {
  const wrapper = getWrapper()
  expect(wrapper.find('p').first()).to.have.text('Step 1')
  expect(wrapper.find('img').first()).to.have.prop('src').equal('foo')
  wrapper.find(NextButton).not('.final-button').first().simulate('click')
  expect(wrapper.find('p').first()).to.have.text('Step 2')
  expect(wrapper.find('img').first()).to.have.prop('src').equal('bar')
})

it('transitions to the previous step when previous is clicked', () => {
  const wrapper = getWrapper().setState({ currentStep: 1 })
  expect(wrapper.find('p').first()).to.have.text('Step 2')
  expect(wrapper.find('img').first()).to.have.prop('src').equal('bar')
  wrapper.find(PreviousButton).first().simulate('click')
  expect(wrapper.find('p').first()).to.have.text('Step 1')
  expect(wrapper.find('img').first()).to.have.prop('src').equal('foo')
})

it('calls onFinalButtonClicked when the final button is clicked', () => {
  const finalClicked = jest.fn()
  const wrapper = getWrapper({ onFinalButtonClicked: finalClicked }).setState({ currentStep: 2 })
  wrapper.find(NextButton).filter('.final-button').first().simulate('click')
  jestExpect(finalClicked).toHaveBeenCalledTimes(1)
})
