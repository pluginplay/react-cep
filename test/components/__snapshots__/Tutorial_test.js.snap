// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  height: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
}

.c6 {
  padding: 4px 8px;
  margin: 10px auto;
  min-width: 100px;
  font-family: "Adobe Clean","Segoe UI","San Francisco",sans-serif;
  color: white;
  background: #191919;
  border: solid 1px #404040;
  font-size: 10pt;
  outline: none;
  box-sizing: border-box;
  -webkit-transition: background 0.1s ease-in-out,border-color 0.1s ease-in-out,-webkit-transform 0.1s ease-in-out, opacity 0.1s ease-in-out;
  -webkit-transition: background 0.1s ease-in-out,border-color 0.1s ease-in-out,transform 0.1s ease-in-out, opacity 0.1s ease-in-out;
  transition: background 0.1s ease-in-out,border-color 0.1s ease-in-out,transform 0.1s ease-in-out, opacity 0.1s ease-in-out;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  cursor: pointer;
  display: inline-block;
}

.c6:not(:disabled):hover,
.c6:not(:disabled):focus,
.c6:not(:disabled):active {
  border-color: #0078d7;
  background: #0c0c0c;
}

.c6:not(:disabled):active {
  -webkit-transform: scale(0.95);
  -ms-transform: scale(0.95);
  transform: scale(0.95);
  background: #262626;
}

.c6:disabled {
  cursor: default;
  opacity: 0.5;
}

.c5 {
  background: #0078d7;
  border: none;
}

.c5 span {
  margin-right: 5px;
}

.c5 img {
  height: 12px;
  margin-left: auto;
}

.c5:hover,
.c5:active,
.c5:focus {
  background: #0078d7;
}

.c5:hover {
  background: #0c0c0c;
}

.c0 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  position: absolute;
  top: 0;
  bottom: 0;
}

.c0 .side {
  width: 50%;
  height: 100%;
  box-sizing: border-box;
}

.c0 .side.image-side {
  background: #191919;
}

.c0 .side.image-side .image-container {
  padding: 15px;
  box-sizing: border-box;
}

.c0 .side.image-side img {
  max-width: 100%;
  max-height: 100%;
}

.c0 .side.copy-side {
  padding: 30px;
}

.c0 .previous-button {
  margin-right: auto;
  margin-left: -8px;
}

.c0 .next-button {
  margin-right: 0;
}

.c0 .final-button {
  font-weight: bold;
  font-size: 14px;
}

.c0 .final-button span {
  margin-right: 10px;
}

.c0 .final-button img {
  height: 14px;
}

.c0 h2 {
  margin: 0;
}

.c0 a {
  color: white;
}

<div
  className="c0 c1"
>
  <div
    className="side image-side"
  >
    <div
      className="image-container c2"
    >
      <img
        src="foo"
      />
    </div>
  </div>
  <div
    className="side copy-side"
  >
    <div
      className="c3"
    >
      <h2>
        Step 1
      </h2>
      <p>
        Step 1
      </p>
      <div
        className="c4"
      >
        <button
          className="next-button c5 c6"
          onClick={[Function]}
        >
          <div
            className="c7"
          >
            <span>
              Step 1 Next
            </span>
            <img
              alt="Next"
              src="test-file-stub"
            />
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
